### ISSUE TITLE
Speed up rigid+affine registration with brain data

### ISSUE THREAD

**Issue by thomshaw92** [ANTsX/ANTs#1498]
Hi experts!
We are trying to speed up our rigid+affine registration steps for faster 'on the go' registration in brain MR images.
I've seen #850, and the 'anatomy of an ants registration call' page, but was wondering you already had solutions for this type of situation that worked well?
Assuming ~1mm iso. I have trialed reducing the number of iterations, number of steps etc - but maybe someone has already done this empirically/heuristically and knows the answer? Of course, we are happy with decreased accuracy/precision with decreased time. 
Thanks for all your help.

**Comment by cookpa** [comment]
Can you give us an example of what you've tried, how long it takes, and how long you'd like it to take?

**Comment by thomshaw92** [comment]
Sure, tried reducing the resolution level steps, the iterations, changing the metric to CC instead of Mattes - but there are so many interacting factors - I wonder what would be the best place to start?
Ideally 30 seconds to a minute? I am guessing writing out the image takes some time too in scripts like SyNQuick.sh  
Thanks again!

**Comment by cookpa** [comment]
Downsampling the images is an option but that's going to eat into your minute. CC will be slower than Mattes. Full resolution iterations can take up > 50% of total run time, so don't do any iterations at -f 1 and fewer at -f 2 if you can.

If you have masks for either image, that can speed things up. If you can preprocess, eg by cropping out irrelevant background (but don't crop right to the edge of the brain), that can help too.
 
Aside from the registration, a few seconds will get spent on other things like reading / writing images, smoothing and downsampling between stages, composing transforms. 

How about something like this:

```
antsRegistration \
  -d 3 
  -r [ $fixed, $moving, 1 ] \
  -t Rigid[ 0.1 ] \
  -m Mattes[ $fixed, $moving, 1, 32, Regular, 0.25 ] \ 
  -f 4x3x2x1 -s 4x3x0x0 -c [ 50x25x10x0, 1e-5, 5 ] \
  -t Affine[ 0.1 ] \
  -m Mattes[ $fixed, $moving, 1, 32, Regular, 0.25 ] \ 
  -f 4x3x2x1 -s 4x3x1x0 -c [ 50x50x20x0, 1e-5, 5 ] \
  -o [ movingToFixed, movingToFixedDeformed.nii.gz, movingToFixedInverseDeformed.nii.gz ] \
  --float 
```

If you have multiple cores and are OK with small variation in results from thread summation, export  `ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=${cores}`, otherwise set `ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1`.

I'm not sure if the subsampled metric `Mattes[ $fixed, $moving, 1, 32, Regular, 0.25 ]` is much faster than the dense version `Mattes[ $fixed, $moving, 1, 32]`. Subsampling too aggressively can cause poor convergence, I wouldn't go lower than 0.125.

**Comment by thomshaw92** [comment]
Thanks so much for all of this info! I will give this a go.
